File: packages/core/src/validate.ts
/**
 * CKM manifest validation against the v2 JSON Schema.
 *
 * @remarks
 * Provides lightweight structural validation without external
 * JSON Schema library dependencies. Validates required fields,
 * types, and enum values as defined in ckm.schema.json.
 *
 * @packageDocumentation
 */

import type { CkmValidationError, CkmValidationResult } from './types.js';

/**
 * Validates a parsed JSON object against the ckm.json v2 schema.
 *
 * @param data - Parsed JSON value.
 * @returns Validation result with JSON pointer paths for errors.
 *
 * @public
 */
export function validateManifest(data: unknown): CkmValidationResult {
  const errors: CkmValidationError[] = [];

  if (typeof data !== 'object' || data === null || Array.isArray(data)) {
    errors.push({ path: '', message: 'Manifest must be an object' });
    return { valid: false, errors };
  }

  const obj = data as Record<string, unknown>;

  // Required top-level fields
  validateRequiredString(obj, 'version', '', errors);
  validateRequiredObject(obj, 'meta', '', errors);
  validateRequiredArray(obj, 'concepts', '', errors);
  validateRequiredArray(obj, 'operations', '', errors);
  validateRequiredArray(obj, 'constraints', '', errors);
  validateRequiredArray(obj, 'workflows', '', errors);
  validateRequiredArray(obj, 'configSchema', '', errors);

  // Meta validation
  if (typeof obj.meta === 'object' && obj.meta !== null) {
    const meta = obj.meta as Record<string, unknown>;
    validateRequiredString(meta, 'project', '/meta', errors);
    validateRequiredString(meta, 'language', '/meta', errors);
    validateRequiredString(meta, 'generator', '/meta', errors);
    validateRequiredString(meta, 'generated', '/meta', errors);
  }

  // Concepts validation
  if (Array.isArray(obj.concepts)) {
    for (let i = 0; i < obj.concepts.length; i++) {
      const path = `/concepts/${i}`;
      const concept = obj.concepts[i] as Record<string, unknown>;
      if (typeof concept !== 'object' || concept === null) {
        errors.push({ path, message: 'Concept must be an object' });
        continue;
      }
      validateRequiredString(concept, 'id', path, errors);
      validateRequiredString(concept, 'name', path, errors);
      validateRequiredString(concept, 'slug', path, errors);
      validateRequiredString(concept, 'what', path, errors);
      validateRequiredArray(concept, 'tags', path, errors);

      if (Array.isArray(concept.properties)) {
        for (let j = 0; j < concept.properties.length; j++) {
          validateProperty(concept.properties[j], `${path}/properties/${j}`, errors);
        }
      }
    }
  }

  // Operations validation
  if (Array.isArray(obj.operations)) {
    for (let i = 0; i < obj.operations.length; i++) {
      const path = `/operations/${i}`;
      const op = obj.operations[i] as Record<string, unknown>;
      if (typeof op !== 'object' || op === null) {
        errors.push({ path, message: 'Operation must be an object' });
        continue;
      }
      validateRequiredString(op, 'id', path, errors);
      validateRequiredString(op, 'name', path, errors);
      validateRequiredString(op, 'what', path, errors);
      validateRequiredArray(op, 'tags', path, errors);
    }
  }

  // Constraints validation
  if (Array.isArray(obj.constraints)) {
    for (let i = 0; i < obj.constraints.length; i++) {
      const path = `/constraints/${i}`;
      const c = obj.constraints[i] as Record<string, unknown>;
      if (typeof c !== 'object' || c === null) {
        errors.push({ path, message: 'Constraint must be an object' });
        continue;
      }
      validateRequiredString(c, 'id', path, errors);
      validateRequiredString(c, 'rule', path, errors);
      validateRequiredString(c, 'enforcedBy', path, errors);
      if (typeof c.severity === 'string') {
        if (!['error', 'warning', 'info'].includes(c.severity)) {
          errors.push({
            path: `${path}/severity`,
            message: `Invalid severity: "${c.severity}". Must be "error", "warning", or "info"`,
          });
        }
      } else {
        errors.push({ path: `${path}/severity`, message: 'Missing required field: severity' });
      }
    }
  }

  // Workflows validation
  if (Array.isArray(obj.workflows)) {
    for (let i = 0; i < obj.workflows.length; i++) {
      const path = `/workflows/${i}`;
      const wf = obj.workflows[i] as Record<string, unknown>;
      if (typeof wf !== 'object' || wf === null) {
        errors.push({ path, message: 'Workflow must be an object' });
        continue;
      }
      validateRequiredString(wf, 'id', path, errors);
      validateRequiredString(wf, 'goal', path, errors);
      validateRequiredArray(wf, 'tags', path, errors);
      if (!Array.isArray(wf.steps) || wf.steps.length === 0) {
        errors.push({ path: `${path}/steps`, message: 'Workflow must have at least one step' });
      }
    }
  }

  // ConfigSchema validation
  if (Array.isArray(obj.configSchema)) {
    for (let i = 0; i < obj.configSchema.length; i++) {
      const path = `/configSchema/${i}`;
      const entry = obj.configSchema[i] as Record<string, unknown>;
      if (typeof entry !== 'object' || entry === null) {
        errors.push({ path, message: 'Config entry must be an object' });
        continue;
      }
      validateRequiredString(entry, 'key', path, errors);
      validateRequiredString(entry, 'description', path, errors);
      if (typeof entry.required !== 'boolean') {
        errors.push({ path: `${path}/required`, message: 'Missing required field: required' });
      }
      validateTypeRef(entry.type, `${path}/type`, errors);
    }
  }

  return { valid: errors.length === 0, errors };
}

// ─── Helpers ─────────────────────────────────────────────────────

const CANONICAL_TYPES = [
  'string',
  'boolean',
  'number',
  'integer',
  'array',
  'object',
  'null',
  'any',
];

function validateRequiredString(
  obj: Record<string, unknown>,
  field: string,
  parentPath: string,
  errors: CkmValidationError[],
): void {
  if (typeof obj[field] !== 'string') {
    errors.push({ path: `${parentPath}/${field}`, message: `Missing required field: ${field}` });
  }
}

function validateRequiredObject(
  obj: Record<string, unknown>,
  field: string,
  parentPath: string,
  errors: CkmValidationError[],
): void {
  if (typeof obj[field] !== 'object' || obj[field] === null) {
    errors.push({ path: `${parentPath}/${field}`, message: `Missing required field: ${field}` });
  }
}

function validateRequiredArray(
  obj: Record<string, unknown>,
  field: string,
  parentPath: string,
  errors: CkmValidationError[],
): void {
  if (!Array.isArray(obj[field])) {
    errors.push({ path: `${parentPath}/${field}`, message: `Missing required field: ${field}` });
  }
}

function validateTypeRef(typeRef: unknown, path: string, errors: CkmValidationError[]): void {
  if (typeof typeRef !== 'object' || typeRef === null) {
    errors.push({ path, message: 'Type reference must be an object' });
    return;
  }
  const ref = typeRef as Record<string, unknown>;
  if (typeof ref.canonical !== 'string' || !CANONICAL_TYPES.includes(ref.canonical)) {
    errors.push({
      path: `${path}/canonical`,
      message: `Invalid canonical type. Must be one of: ${CANONICAL_TYPES.join(', ')}`,
    });
  }
}

function validateProperty(prop: unknown, path: string, errors: CkmValidationError[]): void {
  if (typeof prop !== 'object' || prop === null) {
    errors.push({ path, message: 'Property must be an object' });
    return;
  }
  const p = prop as Record<string, unknown>;
  validateRequiredString(p, 'name', path, errors);
  validateRequiredString(p, 'description', path, errors);
  if (typeof p.required !== 'boolean') {
    errors.push({ path: `${path}/required`, message: 'Missing required field: required' });
  }
  validateTypeRef(p.type, `${path}/type`, errors);
}

File: packages/rust-core/src/validate.rs
//! CKM manifest validation against the v2 schema.
//!
//! Provides lightweight structural validation without external JSON Schema
//! library dependencies. Validates required fields, types, and enum values.

use serde_json::Value;

use crate::types::{CkmValidationError, CkmValidationResult};

const CANONICAL_TYPES: &[&str] = &[
    "string", "boolean", "number", "integer", "array", "object", "null", "any",
];

/// Validates a parsed JSON object against the ckm.json v2 schema.
///
/// Returns `{ valid: true, errors: [] }` for valid manifests.
/// Returns `{ valid: false, errors: [...] }` with JSON pointer paths for invalid manifests.
/// v1 manifests will fail validation (they lack required v2 fields).
pub fn validate_manifest(data: &Value) -> CkmValidationResult {
    let mut errors: Vec<CkmValidationError> = Vec::new();

    let obj = match data.as_object() {
        Some(obj) => obj,
        None => {
            errors.push(CkmValidationError {
                path: String::new(),
                message: "Manifest must be an object".to_string(),
            });
            return CkmValidationResult {
                valid: false,
                errors,
            };
        }
    };

    // Required top-level fields
    validate_required_string(obj, "version", "", &mut errors);
    validate_required_object(obj, "meta", "", &mut errors);
    validate_required_array(obj, "concepts", "", &mut errors);
    validate_required_array(obj, "operations", "", &mut errors);
    validate_required_array(obj, "constraints", "", &mut errors);
    validate_required_array(obj, "workflows", "", &mut errors);
    validate_required_array(obj, "configSchema", "", &mut errors);

    // Meta validation
    if let Some(Value::Object(meta)) = obj.get("meta") {
        validate_required_string(meta, "project", "/meta", &mut errors);
        validate_required_string(meta, "language", "/meta", &mut errors);
        validate_required_string(meta, "generator", "/meta", &mut errors);
        validate_required_string(meta, "generated", "/meta", &mut errors);
    }

    // Concepts validation
    if let Some(Value::Array(concepts)) = obj.get("concepts") {
        for (i, concept) in concepts.iter().enumerate() {
            let path = format!("/concepts/{}", i);
            if let Some(c) = concept.as_object() {
                validate_required_string(c, "id", &path, &mut errors);
                validate_required_string(c, "name", &path, &mut errors);
                validate_required_string(c, "slug", &path, &mut errors);
                validate_required_string(c, "what", &path, &mut errors);
                validate_required_array(c, "tags", &path, &mut errors);

                if let Some(Value::Array(props)) = c.get("properties") {
                    for (j, prop) in props.iter().enumerate() {
                        validate_property(prop, &format!("{}/properties/{}", path, j), &mut errors);
                    }
                }
            } else {
                errors.push(CkmValidationError {
                    path,
                    message: "Concept must be an object".to_string(),
                });
            }
        }
    }

    // Operations validation
    if let Some(Value::Array(operations)) = obj.get("operations") {
        for (i, operation) in operations.iter().enumerate() {
            let path = format!("/operations/{}", i);
            if let Some(op) = operation.as_object() {
                validate_required_string(op, "id", &path, &mut errors);
                validate_required_string(op, "name", &path, &mut errors);
                validate_required_string(op, "what", &path, &mut errors);
                validate_required_array(op, "tags", &path, &mut errors);
            } else {
                errors.push(CkmValidationError {
                    path,
                    message: "Operation must be an object".to_string(),
                });
            }
        }
    }

    // Constraints validation
    if let Some(Value::Array(constraints)) = obj.get("constraints") {
        for (i, constraint) in constraints.iter().enumerate() {
            let path = format!("/constraints/{}", i);
            if let Some(c) = constraint.as_object() {
                validate_required_string(c, "id", &path, &mut errors);
                validate_required_string(c, "rule", &path, &mut errors);
                validate_required_string(c, "enforcedBy", &path, &mut errors);
                if let Some(Value::String(severity)) = c.get("severity") {
                    if !["error", "warning", "info"].contains(&severity.as_str()) {
                        errors.push(CkmValidationError {
                            path: format!("{}/severity", path),
                            message: format!(
                                "Invalid severity: \"{}\". Must be \"error\", \"warning\", or \"info\"",
                                severity
                            ),
                        });
                    }
                } else {
                    errors.push(CkmValidationError {
                        path: format!("{}/severity", path),
                        message: "Missing required field: severity".to_string(),
                    });
                }
            } else {
                errors.push(CkmValidationError {
                    path,
                    message: "Constraint must be an object".to_string(),
                });
            }
        }
    }

    // Workflows validation
    if let Some(Value::Array(workflows)) = obj.get("workflows") {
        for (i, workflow) in workflows.iter().enumerate() {
            let path = format!("/workflows/{}", i);
            if let Some(wf) = workflow.as_object() {
                validate_required_string(wf, "id", &path, &mut errors);
                validate_required_string(wf, "goal", &path, &mut errors);
                validate_required_array(wf, "tags", &path, &mut errors);
                match wf.get("steps") {
                    Some(Value::Array(steps)) if !steps.is_empty() => {}
                    _ => {
                        errors.push(CkmValidationError {
                            path: format!("{}/steps", path),
                            message: "Workflow must have at least one step".to_string(),
                        });
                    }
                }
            } else {
                errors.push(CkmValidationError {
                    path,
                    message: "Workflow must be an object".to_string(),
                });
            }
        }
    }

    // ConfigSchema validation
    if let Some(Value::Array(config_schema)) = obj.get("configSchema") {
        for (i, entry) in config_schema.iter().enumerate() {
            let path = format!("/configSchema/{}", i);
            if let Some(e) = entry.as_object() {
                validate_required_string(e, "key", &path, &mut errors);
                validate_required_string(e, "description", &path, &mut errors);
                if e.get("required").and_then(Value::as_bool).is_none() {
                    errors.push(CkmValidationError {
                        path: format!("{}/required", path),
                        message: "Missing required field: required".to_string(),
                    });
                }
                validate_type_ref(e.get("type"), &format!("{}/type", path), &mut errors);
            } else {
                errors.push(CkmValidationError {
                    path,
                    message: "Config entry must be an object".to_string(),
                });
            }
        }
    }

    CkmValidationResult {
        valid: errors.is_empty(),
        errors,
    }
}

// ─── Helpers ────────────────────────────────────────────────────────────

fn validate_required_string(
    obj: &serde_json::Map<String, Value>,
    field: &str,
    parent_path: &str,
    errors: &mut Vec<CkmValidationError>,
) {
    if obj.get(field).and_then(Value::as_str).is_none() {
        errors.push(CkmValidationError {
            path: format!("{}/{}", parent_path, field),
            message: format!("Missing required field: {}", field),
        });
    }
}

fn validate_required_object(
    obj: &serde_json::Map<String, Value>,
    field: &str,
    parent_path: &str,
    errors: &mut Vec<CkmValidationError>,
) {
    if obj.get(field).and_then(Value::as_object).is_none() {
        errors.push(CkmValidationError {
            path: format!("{}/{}", parent_path, field),
            message: format!("Missing required field: {}", field),
        });
    }
}

fn validate_required_array(
    obj: &serde_json::Map<String, Value>,
    field: &str,
    parent_path: &str,
    errors: &mut Vec<CkmValidationError>,
) {
    if obj.get(field).and_then(Value::as_array).is_none() {
        errors.push(CkmValidationError {
            path: format!("{}/{}", parent_path, field),
            message: format!("Missing required field: {}", field),
        });
    }
}

fn validate_type_ref(type_ref: Option<&Value>, path: &str, errors: &mut Vec<CkmValidationError>) {
    match type_ref.and_then(Value::as_object) {
        Some(ref_obj) => {
            if let Some(Value::String(canonical)) = ref_obj.get("canonical") {
                if !CANONICAL_TYPES.contains(&canonical.as_str()) {
                    errors.push(CkmValidationError {
                        path: format!("{}/canonical", path),
                        message: format!(
                            "Invalid canonical type. Must be one of: {}",
                            CANONICAL_TYPES.join(", ")
                        ),
                    });
                }
            } else {
                errors.push(CkmValidationError {
                    path: format!("{}/canonical", path),
                    message: format!(
                        "Invalid canonical type. Must be one of: {}",
                        CANONICAL_TYPES.join(", ")
                    ),
                });
            }
        }
        None => {
            errors.push(CkmValidationError {
                path: path.to_string(),
                message: "Type reference must be an object".to_string(),
            });
        }
    }
}

fn validate_property(prop: &Value, path: &str, errors: &mut Vec<CkmValidationError>) {
    match prop.as_object() {
        Some(p) => {
            validate_required_string(p, "name", path, errors);
            validate_required_string(p, "description", path, errors);
            if p.get("required").and_then(Value::as_bool).is_none() {
                errors.push(CkmValidationError {
                    path: format!("{}/required", path),
                    message: "Missing required field: required".to_string(),
                });
            }
            validate_type_ref(p.get("type"), &format!("{}/type", path), errors);
        }
        None => {
            errors.push(CkmValidationError {
                path: path.to_string(),
                message: "Property must be an object".to_string(),
            });
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_manifest() {
        let data: Value = serde_json::json!({
            "$schema": "https://ckm.dev/schemas/v2.json",
            "version": "2.0.0",
            "meta": {
                "project": "test",
                "language": "typescript",
                "generator": "test",
                "generated": "2026-01-01T00:00:00.000Z"
            },
            "concepts": [],
            "operations": [],
            "constraints": [],
            "workflows": [],
            "configSchema": []
        });
        let result = validate_manifest(&data);
        assert!(result.valid);
        assert!(result.errors.is_empty());
    }

    #[test]
    fn test_invalid_manifest_missing_meta() {
        let data: Value = serde_json::json!({
            "version": "2.0.0",
            "concepts": [],
            "operations": [],
            "constraints": [],
            "workflows": [],
            "configSchema": []
        });
        let result = validate_manifest(&data);
        assert!(!result.valid);
        assert!(result.errors.iter().any(|e| e.path == "/meta"));
    }

    #[test]
    fn test_invalid_manifest_not_object() {
        let data: Value = serde_json::json!(42);
        let result = validate_manifest(&data);
        assert!(!result.valid);
        assert_eq!(result.errors[0].message, "Manifest must be an object");
    }
}
